lib/fibonacci.rb

```diff
@@ -1,8 +1,44 @@
 # Improved Fibonacci
 
-# Time Complexity - ?
-# Space Complexity - ?  (should be O(n))
+## Improved Fibonacci
+
+# Earlier you saw how an array could be used to store Fibonacci numbers resulting in a time-complexity reduction from O(2<sup>n</sup>) to O(n).  Now we will take this a set further, because to find a given Fibonacci number, you only need to find the previous two numbers.  
+
+# Reminder:
+
+# Fib(0) = 0
+# Fib(1) = 1
+
+# Fib(n) = Fib(n-2) + Fib(n-1), for all n >= 2
+
+# Restrictions:
+
+#   - You cannot use a loop, use recursion.
+#   - Your solution should be O(n) space complexity, due to the call stack.
+
+
+
+# Recursive Solution from class
+
+# Time Complexity - O(2^n)
+# Space Complexity - O(n)
 # Hint, you may want a recursive helper method
 def fibonacci(n)
```

Comment:
Your time complexity is actually O(n) because you store previous problems in the solutions array.